xwayland/selection/outgoing.c

```diff
@@ -312,14 +316,34 @@ static void xwm_selection_send_data(struct wlr_xwm_selection *selection,
 	transfer->source_fd = p[0];
 
 	wlr_log(WLR_DEBUG, "Sending Wayland selection %u to Xwayland window with "
-		"MIME type %s, target %u", req->target, mime_type, req->target);
+		"MIME type %s, target %u, transfer %p", req->target, mime_type,
+		req->target, transfer);
 	xwm_selection_source_send(selection, mime_type, p[1]);
 
+	// It seems that if we ever try to reply to a selection request after
+	// another has been sent by the same requestor, the requestor never reads
+	// from it. It appears to only ever read from the latest, so purge stale
+	// transfers to prevent clipboard hangs.
+	struct wlr_xwm_selection_transfer *outgoing, *tmp;
+	wl_list_for_each_safe(outgoing, tmp, &selection->outgoing, outgoing_link) {
+		if (transfer->request.requestor == outgoing->request.requestor) {
```

Comment:
Side note: maybe it would be a good idea to filter by other `xcb_selection_request_event_t` fields if this turns out not being narrow enough.